sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -579,23 +593,52 @@ case class AdaptiveSparkPlanExec(
         allChildStagesMaterialized = false,
         newStages = Seq(newStage))
 
-    case q: QueryStageExec =>
+    case q: QueryStageExec if q ne currentPhysicalPlan =>
       assertStageNotFailed(q)
       CreateStageResult(newPlan = q,
         allChildStagesMaterialized = q.isMaterialized, newStages = Seq.empty)
 
     case _ =>
-      if (plan.children.isEmpty) {
-        CreateStageResult(newPlan = plan, allChildStagesMaterialized = true, newStages = Seq.empty)
+      // This includes cases where root node is QueryStageExec
+      val results = plan.children.map(createQueryStages(resultHandler, _, firstRun))
+      val allChildStagesMaterialized = results.forall(_.allChildStagesMaterialized)
+      val newStages = results.flatMap(_.newStages)
+      if ((plan eq currentPhysicalPlan) && allChildStagesMaterialized && newStages.isEmpty) {
+        // Create result stage if the plan is root plan and all children stages are materialized
+        // and no new stages are created.
+        val resultStage = createResultQueryStage(resultHandler, plan)
+        stagesToReplace.append(resultStage)
+        CreateStageResult(
+          newPlan = resultStage,
+          allChildStagesMaterialized = false,
+          newStages = Seq(resultStage))
       } else {
-        val results = plan.children.map(createQueryStages)
+        val newPlan = if (results.isEmpty) {
+          plan
+        } else {
+          plan.withNewChildren(results.map(_.newPlan))
+        }
         CreateStageResult(
-          newPlan = plan.withNewChildren(results.map(_.newPlan)),
-          allChildStagesMaterialized = results.forall(_.allChildStagesMaterialized),
-          newStages = results.flatMap(_.newStages))
+          newPlan = newPlan,
+          allChildStagesMaterialized = allChildStagesMaterialized,
+          newStages = newStages)
       }
   }
 
+  private def createResultQueryStage(
```

Comment:
```suggestion
  private def newResultQueryStage(
```

To follow the name of `newQueryStage`